### DISCUSSION TITLE
F1 null

### DISCUSSION THREAD

**Discussion by asoxien** [azavea/raster-vision#1504] [Q&A]
Hi everyone;
After training object detection I get f1 null !!!
**"overall": [
{
"class_id": 0,
"class_name": "car",
"gt_count": 2414.0,
"pred_count": 0.0,
"count_error": 2414.0,
"metrics": {
"recall": 0.0,
"precision": null,
"f1": null,
"sensitivity": 0.0,
"specificity": null
},
"true_pos": 0.0,
"false_pos": 0.0,
"false_neg": 2414.0
},**

I used different backbone but the same result 
The used configuration for training:
**model=ObjectDetectionModelConfig(backbone=<Backbone.resnet50: 'resnet50'>, pretrained=True, init_weights=None, load_strict=True, external_def=None) solver=SolverConfig(lr=0.0001, num_epochs=100, test_num_epochs=2, test_batch_sz=4, overfit_num_steps=1, sync_interval=1, batch_sz=200, one_cycle=True, multi_stage=[], class_loss_weights=None, ignore_class_index=None, external_loss_def=None) data=ObjectDetectionGeoDataConfig(scene_dataset='<93 train_scenes, 41 validation_scenes, 0 test_scenes>', window_opts="method=<GeoDataWindowMethod.random: 'random'> size=412 stride=None padding=None size_lims=(412, 413) h_lims=None w_lims=None max_windows=500 max_sample_attempts=100 efficient_aoi_sampling=True ioa_thresh=0.75 clip=True neg_ratio=None neg_ioa_thresh=0.2") predict_mode=False test_mode=False overfit_mode=False eval_train=False save_model_bundle=True log_tensorboard=False run_tensorboard=False output_uri='/opt/src/output/model-object_detector_test/train'**

**Comment by NoaMills-USDA** [comment]
This is because the model predicts no true positives and no false positives, so when calculating precision and f1, we are dividing by zero. The formula for precision is TP/(TP+FP), which in this case would be 0/0 which is indeterminate, thus listed as null. The f1 value is based on both the values for recall and precision, and since precision is null, we can't find a value for f1 so f1 is then null.

Your model is predicting everything to be "not a car".

**Comment by AdeelH** [comment]
What dataset are you training on?

**Reply by asoxien** [reply]
I use parts of Pleiade images that I annotated by my self

**Reply by asoxien** [reply]
I used also for test this two images with labels:
training:
https://s3.amazonaws.com/azavea-research-public-data/raster-vision/examples/spacenet/RGB-PanSharpen_AOI_2_Vegas_img25.tif
https://s3.amazonaws.com/azavea-research-public-data/raster-vision/examples/spacenet/buildings_AOI_2_Vegas_img25.geojson
validation:
https://s3.amazonaws.com/azavea-research-public-data/raster-vision/examples/spacenet/RGB-PanSharpen_AOI_2_Vegas_img205.tif
https://s3.amazonaws.com/azavea-research-public-data/raster-vision/examples/spacenet/buildings_AOI_2_Vegas_img205.geojson
but I get the same results. I think there are some bugs !!

**Reply by AdeelH** [reply]
- Are you using 0.13 or the master branch?
- Is the training data being read correctly? Check the images in `train/dataloaders` to see if they look as expected.
- What are the metrics like during training? Do they improve from one epoch to the next?

**Reply by asoxien** [reply]
Hi Adeel;
- I use master branch the latest version
- I train the model for 10 epochs, the training metrics seems to improve as seen in the following log.csv:
epoch,train_time,valid_time,**train_loss**,map,map50
0,0:01:11.982944,0:00:11.204795,**0.009905728839692615**,0.0,0.0
1,0:01:07.497235,0:00:11.558573,**2.577289706096053e-07**,0.0,0.0
2,0:01:06.602937,0:00:09.931241,**4.0624831204435654e-08**,0.0,0.0
3,0:01:06.702624,0:00:09.547714,**1.0266156182513528e-08**,0.0,0.0
4,0:01:06.640584,0:00:09.476570,**3.6507134652873945e-09**,0.0,0.0
5,0:01:06.519504,0:00:09.431838,**1.796299026561125e-09**,0.0,0.0
6,0:01:13.035444,0:00:10.768223,**7.382755570905006e-10**,0.0,0.0
7,0:01:05.552953,0:00:09.601797,**3.5578371656204884e-10**,0.0,0.0
8,0:01:04.859484,0:00:09.290950,**2.380401535143998e-10**,0.0,0.0
9,0:01:04.817704,0:00:09.252162,**1.2883249599413136e-10**,0.0,0.0

- the valid and train png are attached I can't tell if they are ok or not !!!:
![train](https://user-images.githubusercontent.com/4633064/196759751-3290442c-7db1-426a-832d-edc60dd09804.png)
![valid](https://user-images.githubusercontent.com/4633064/196759759-c0407473-f54d-439f-a4c6-1c4e7085640e.png)

I can share with you anything (images/labels/pipefile-config ...) just ask if necessary 
thank you

**Reply by asoxien** [reply]
hi;
I use also 0.13.1 with the same metadata and same configuration except negative ratio that is not accepted in the master version.
the training data read correctly I think the problem in the master version is the reading of the training data

**Reply by AdeelH** [reply]
>I use also 0.13.1 with the same metadata and same configuration except negative ratio that is not accepted in the master version.
the training data read correctly I think the problem in the master version is the reading of the training data

This is interesting. Can you share the dataloader images when using 0.13.1? Also what `neg_ratio` are you using?

**Reply by asoxien** [reply]
the "neg_ratio": 0.95 and for the dataloader images are as follow: 
by the way even if "test_scenes": [] means no data used for test I get in the dataloader test.png (this problem doesn't appear in the master branch). 
validation png:
![valid](https://user-images.githubusercontent.com/4633064/197546726-b64e89cf-32fe-4b61-8e08-197e2bcef7f8.png)
train png:
![train](https://user-images.githubusercontent.com/4633064/197545928-d1ab8471-837c-494e-9929-3767b46a65a4.png)
test png:
![test](https://user-images.githubusercontent.com/4633064/197545361-77d4370f-ade8-4ab1-92f5-fda01961f248.png)

**Reply by AdeelH** [reply]
So, with the master branch, you only got 2 items in each image even though your batch size was 200? Or were those earlier images generated with a different batch size?

>by the way even if "test_scenes": [] means no data used for test I get in the dataloader test.png (this problem doesn't appear in the master branch).

Yeah, the test set is equal to the validation set in that case. And yes, this is not the case on master.

**Reply by AdeelH** [reply]
I would also recommend using a much smaller chip size, since the image resolution is not very high and you can barely make out cars in the chips.

**Reply by asoxien** [reply]
Yes exactly  the same images used in the two cases with the same value "batch_sz": 200.
Only the neg_ratio is different

**Reply by asoxien** [reply]
by the way I set again the negative ratio = 0.95 in the master test and I get this error:
ValueError: neg_ratio specified, but no bboxes found in scene.
Means the program doesn't read bboxes correctly in the master example. example of labels file:
{
"type": "FeatureCollection",
"name": "train-9",
"crs": { "type": "name", "properties": { "name": "urn:ogc:def:crs:OGC:1.3:CRS84" } },
"bbox": [ 144.946730926839, -37.8476248101864, 144.949048894182, -37.8457556247624 ],                                              
"features": [
{ "type": "Feature", "properties": { "id": null }, "bbox": [ 144.948977131594717, -37.846744460039751, 144.949027716548642, -37.846684703617079 ], "geometry": { "type": "MultiPolygon", "coordinates": [ [ [ [ 144.949023021898228, -37.846684703617079 ], [ 144.949027716548642, -37.846742099794263 ], [ 144.948981826209547, -37.846744460039751 ], [ 144.948977131594717, -37.846687063861779 ], [ 144.949023021898228, -37.846684703617079 ] ] ] ] } },
{ "type": "Feature", "properties": { "id": null }, "bbox": [ 144.948534112757471, -37.846720526111909, 144.94862245020326, -37.846679918564412 ], "geometry": { "type": "MultiPolygon", "coordinates": [ [ [ [ 144.94862245020326, -37.846709495689446 ], [ 144.94854043916547, -37.846720526111909 ], [ 144.948534112757471, -37.846690948983436 ], [ 144.948616123763031, -37.846679918564412 ], [ 144.94862245020326, -37.846709495689446 ] ] ] ] } }
]
}

**Reply by AdeelH** [reply]
On `master`, can you try the following:
- start docker with `docker/run --aws --gpu --jupyter-lab`
- wait for a few seconds for the Jupyter Lab server to start
- open a new notebook
- try running this example, with the URIs and `ClassConfig` changed to match your dataset: https://docs.rastervision.io/en/latest/tutorials/reading_labels.html#Object-Detection---ObjectDetectionLabelSource
    - also change the indices in these lines below to something more suited to your raster's size:
        ```python
        chip = raster_source[:1000:2, :1000:2]
        bboxes, _, _ = label_source[:1000:2, :1000:2]
        ```
- see if the bounding boxes show up correctly

**Reply by lewfish** [reply]
I have the same problem on the `master` branch with the `cowc-potsdam` example. I'm going to look into it today.

**Reply by lewfish** [reply]
It looks like the chips are being read correctly and predictions look correct too.

**Reply by asoxien** [reply]
Hi Adeel;

the access to images is denied I think the link is deprecated 
**img_uri = 'https://s3.amazonaws.com://raster-vision-raw-data/isprs-potsdam/4_Ortho_RGBIR/top_potsdam_2_10_RGBIR.tif'**
<Error>
<Code>AccessDenied</Code>
<Message>Access Denied</Message>
<RequestId>FZCRX76HR1E07NBW</RequestId>
<HostId>JvoVPpD52+Bwnn7puhHNEtv0w3qmUxAMUWCqmSst3zDW/8gSihSKMgcvgT1+3cFhjItshuBliP8=</HostId>
</Error>

**Reply by AdeelH** [reply]
You should replace that URI with the path of your image. 

That link does need to be updated though. Thanks for catching that!

**Reply by lewfish** [reply]
Actually, the predictions appear to be good in the training command, but the output of the predict command is empty JSON files.

**Reply by asoxien** [reply]
as you can see the bounding boxes show up correctly:
![Selection_001](https://user-images.githubusercontent.com/4633064/198069609-450e13ca-cebf-4de3-9681-69169e19896a.png)

**Reply by asoxien** [reply]
I found something that can help when I put negative ratio in master it gets me this error:
ValueError: neg_ratio specified, but no bboxes found in scene.

**Reply by asoxien** [reply]
Find the result of cowc_potsdam to have a clear idea:
[cowc.zip](https://github.com/azavea/raster-vision/files/9880318/cowc.zip)

**Comment by lewfish** [comment]
I opened an issue for this and have narrowed down the bug, but haven't figured it out yet. See https://github.com/azavea/raster-vision/issues/1517

**Reply by asoxien** [reply]
Hi;
The #1519 corrected the problem thank you.
But I was obliged to change the files in an old version before #1501 because this last (the last master version) throws me this error **NameError: name 'ObjectDetectionConfig' is not defined**

**Reply by lewfish** [reply]
PR #1519 has a fix for that problem too now.
